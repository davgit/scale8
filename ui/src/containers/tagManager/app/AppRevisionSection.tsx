import { FC } from 'react';
import { useQuery } from '@apollo/client';
import TagIcon from '../../../components/atoms/Icons/TagIcon';
import PlatformRevisionIcon from '../../../components/atoms/Icons/PlatformRevisionIcon';
import NavAppRevisionQuery from '../../../gql/queries/NavAppRevisionQuery';
import { NavAppRevision } from '../../../gql/generated/NavAppRevision';
import {
    buildAppButtonProps,
    buildAppRevisionButtonProps,
    buildOrgButtonProps,
    buildTagManagerButtonProps,
} from '../../../utils/BreadcrumbButtonsUtils';
import { Section, SectionProps } from '../../abstractions/Section';
import ActionIcon from '../../../components/atoms/Icons/ActionIcon';
import TriggerIcon from '../../../components/atoms/Icons/TriggerIcon';
import { buildAppRevisionBreadcrumbActions } from '../../../utils/BuildAppRevisionBreadcrumbActions';
import { extractPermissionsFromOrgUser } from '../../../context/OrgUserReducer';
import { SectionKey } from '../../SectionsDetails';
import { useLoggedInState } from '../../../context/AppContext';
import { useRouter } from 'next/router';
import { ChildrenAndIdProps } from '../../../types/props/ChildrenAndIdProps';
import { toAppRevision } from '../../../utils/NavigationPaths';
import { analyticsEnabled } from '../../../utils/AnalyticsUtils';

const AppRevisionSection: FC<ChildrenAndIdProps> = (props: ChildrenAndIdProps) => {
    const router = useRouter();

    const { id, children } = props;

    const { orgUserState, templateInteractions } = useLoggedInState();
    const { ask, dispatchDialogAction, setRefreshCurrentPage } = templateInteractions;

    const currentOrgPermissions = extractPermissionsFromOrgUser(orgUserState);

    const sectionProps: SectionProps<NavAppRevision> = {
        children,
        sectionKey: SectionKey.appRevision,
        queryResult: useQuery<NavAppRevision>(NavAppRevisionQuery, {
            variables: { id },
        }),
        initContext: (data) => {
            templateInteractions.setSectionHasAnalytics(analyticsEnabled(data.getRevision.app));
        },
        buildButtonsProps: (data) => [
            buildOrgButtonProps(
                router,
                data.me.orgs,
                data.getRevision.app.tag_manager_account.org.id,
                data.getRevision.app.tag_manager_account.org.name,
            ),
            buildTagManagerButtonProps(
                router,
                data.getRevision.app.tag_manager_account.id,
                data.getRevision.app.tag_manager_account.org.data_manager_account?.id ?? '',
            ),
            buildAppButtonProps(
                router,
                data.getRevision.app.tag_manager_account.apps,
                data.getRevision.app.id,
                data.getRevision.app.name,
            ),
            buildAppRevisionButtonProps(
                router,
                data.getRevision.app.revisions,
                id,
                data.getRevision.name,
                true,
            ),
        ],
        buildMenuItemsProps: () => [
            {
                icon: () => <TagIcon />,
                label: 'Tags',
                link: toAppRevision({ id }, 'tags'),
            },
            {
                icon: () => <TriggerIcon />,
                label: 'Global Triggers',
                link: toAppRevision({ id }, 'global-triggers'),
            },
            {
                icon: () => <ActionIcon />,
                label: 'Global Actions',
                link: toAppRevision({ id }, 'global-actions'),
            },
            {
                icon: () => <PlatformRevisionIcon />,
                label: 'Platform Revisions',
                link: toAppRevision({ id }, 'app-platform-revisions'),
            },
        ],
        extractOrgUserDetails: (data) => data.getRevision.app.tag_manager_account.org,
        buildBreadcrumbActions: (data) =>
            buildAppRevisionBreadcrumbActions(
                id,
                data.getRevision.name,
                router,
                currentOrgPermissions,
                dispatchDialogAction,
                setRefreshCurrentPage,
                ask,
                data.getRevision.locked,
            ),
        accountExpireIn: orgUserState?.tagManagerAccount?.trialExpiration ?? undefined,
        accountIsTrial: orgUserState?.tagManagerAccount?.isTrial ?? undefined,
    };

    return <Section<NavAppRevision> {...sectionProps} />;
};

export default AppRevisionSection;
