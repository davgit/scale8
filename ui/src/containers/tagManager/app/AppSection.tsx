import { FC } from 'react';
import { useQuery } from '@apollo/client';
import { NavApp } from '../../../gql/generated/NavApp';
import NavAppQuery from '../../../gql/queries/NavAppQuery';
import RevisionIcon from '../../../components/atoms/Icons/RevisionIcon';
import EnvironmentIcon from '../../../components/atoms/Icons/EnvironmentIcon';
import PlatformIcon from '../../../components/atoms/Icons/PlatformIcon';
import {
    buildAppButtonProps,
    buildOrgButtonProps,
    buildTagManagerButtonProps,
} from '../../../utils/BreadcrumbButtonsUtils';
import { Section, SectionProps } from '../../abstractions/Section';
import AppAnalyticsIcon from '../../../components/atoms/Icons/AppAnalyticsIcon';
import { SectionKey } from '../../SectionsDetails';
import { useLoggedInState } from '../../../context/AppContext';
import { useRouter } from 'next/router';
import { ChildrenAndIdProps } from '../../../types/props/ChildrenAndIdProps';
import { toApp } from '../../../utils/NavigationPaths';
import { analyticsEnabled } from '../../../utils/AnalyticsUtils';
import { SideMenuButtonProps } from '../../../components/molecules/SideMenuButton';

const AppSection: FC<ChildrenAndIdProps> = (props: ChildrenAndIdProps) => {
    const router = useRouter();

    const { id, children } = props;

    const { orgUserState, templateInteractions } = useLoggedInState();

    const sectionProps: SectionProps<NavApp> = {
        children,
        sectionKey: SectionKey.app,
        queryResult: useQuery<NavApp>(NavAppQuery, {
            variables: { id },
        }),
        initContext: (data) => {
            templateInteractions.setSectionHasAnalytics(analyticsEnabled(data.getApp));
        },
        buildButtonsProps: (data) => [
            buildOrgButtonProps(
                router,
                data.me.orgs,
                data.getApp.tag_manager_account.org.id,
                data.getApp.tag_manager_account.org.name,
            ),
            buildTagManagerButtonProps(
                router,
                data.getApp.tag_manager_account.id,
                data.getApp.tag_manager_account.org.data_manager_account?.id ?? '',
            ),
            buildAppButtonProps(
                router,
                data.getApp.tag_manager_account.apps,
                id,
                data.getApp.name,
                true,
            ),
        ],
        buildMenuItemsProps: (_, data) => [
            ...((analyticsEnabled(data.getApp)
                ? [
                      {
                          icon: <AppAnalyticsIcon />,
                          label: 'Analytics',
                          link: toApp({ id }, 'analytics'),
                      },
                  ]
                : []) as SideMenuButtonProps[]),
            {
                icon: <RevisionIcon />,
                label: 'Revisions',
                link: toApp({ id }, 'revisions'),
            },
            {
                icon: <EnvironmentIcon />,
                label: 'Environments',
                link: toApp({ id }, 'environments'),
            },
            {
                icon: <PlatformIcon />,
                label: 'Installed Platforms',
                link: toApp({ id }, 'platforms'),
            },
        ],
        extractOrgUserDetails: (data) => data.getApp.tag_manager_account.org,
        accountExpireIn: orgUserState?.tagManagerAccount?.trialExpiration ?? undefined,
        accountIsTrial: orgUserState?.tagManagerAccount?.isTrial ?? undefined,
    };

    return <Section<NavApp> {...sectionProps} />;
};

export default AppSection;
