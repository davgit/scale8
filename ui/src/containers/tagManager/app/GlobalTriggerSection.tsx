import { FC } from 'react';
import { useQuery } from '@apollo/client';
import NavGlobalTriggerQuery from '../../../gql/queries/NavGlobalTriggerQuery';
import { NavGlobalTrigger } from '../../../gql/generated/NavGlobalTrigger';
import {
    buildAppButtonProps,
    buildAppRevisionButtonProps,
    buildGlobalTriggerButtonProps,
    buildOrgButtonProps,
    buildTagManagerButtonProps,
} from '../../../utils/BreadcrumbButtonsUtils';
import { Section, SectionProps } from '../../abstractions/Section';
import { buildAppRevisionBreadcrumbActions } from '../../../utils/BuildAppRevisionBreadcrumbActions';
import { extractPermissionsFromOrgUser } from '../../../context/OrgUserReducer';
import { SectionKey } from '../../SectionsDetails';
import { useLoggedInState } from '../../../context/AppContext';
import { useRouter } from 'next/router';
import { ChildrenAndIdProps } from '../../../types/props/ChildrenAndIdProps';
import { analyticsEnabled } from '../../../utils/AnalyticsUtils';

const GlobalTriggerSection: FC<ChildrenAndIdProps> = (props: ChildrenAndIdProps) => {
    const router = useRouter();

    const { id, children } = props;

    const { orgUserState, templateInteractions } = useLoggedInState();

    const { ask, dispatchDialogAction, setRefreshCurrentPage } =
        templateInteractions;
    const currentOrgPermissions = extractPermissionsFromOrgUser(orgUserState);

    const sectionProps: SectionProps<NavGlobalTrigger> = {
        children,
        sectionKey: SectionKey.globalTrigger,
        queryResult: useQuery<NavGlobalTrigger>(NavGlobalTriggerQuery, {
            variables: { id },
        }),
        initContext: (data) => {
            templateInteractions.setSectionHasAnalytics(
                analyticsEnabled(data.getTrigger.revision.app),
            );
        },
        buildButtonsProps: (data) => [
            buildOrgButtonProps(
                router,
                data.me.orgs,
                data.getTrigger.revision.app.tag_manager_account.org.id,
                data.getTrigger.revision.app.tag_manager_account.org.name,
            ),
            buildTagManagerButtonProps(
                router,
                data.getTrigger.revision.app.tag_manager_account.id,
                data.getTrigger.revision.app.tag_manager_account.org.data_manager_account?.id ?? '',
            ),
            buildAppButtonProps(
                router,
                data.getTrigger.revision.app.tag_manager_account.apps,
                data.getTrigger.revision.app.id,
                data.getTrigger.revision.app.name,
            ),
            buildAppRevisionButtonProps(
                router,
                data.getTrigger.revision.app.revisions,
                data.getTrigger.revision.id,
                data.getTrigger.revision.name,
            ),
            buildGlobalTriggerButtonProps(
                router,
                data.getTrigger.revision.global_triggers,
                id,
                data.getTrigger.name,
                true,
            ),
        ],
        buildMenuItemsProps: () => [],
        extractOrgUserDetails: (data) => data.getTrigger.revision.app.tag_manager_account.org,
        buildBreadcrumbActions: (data) =>
            buildAppRevisionBreadcrumbActions(
                data.getTrigger.revision.id,
                data.getTrigger.revision.name,
                router,
                currentOrgPermissions,
                dispatchDialogAction,
                setRefreshCurrentPage,
                ask,
                data.getTrigger.revision.locked,
            ),
        accountExpireIn: orgUserState?.tagManagerAccount?.trialExpiration ?? undefined,
        accountIsTrial: orgUserState?.tagManagerAccount?.isTrial ?? undefined,
    };

    return <Section<NavGlobalTrigger> {...sectionProps} />;
};

export default GlobalTriggerSection;
