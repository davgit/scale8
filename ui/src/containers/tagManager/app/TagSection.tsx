import { FC } from 'react';
import { useQuery } from '@apollo/client';
import NavTagQuery from '../../../gql/queries/NavTagQuery';
import { NavTag } from '../../../gql/generated/NavTag';
import {
    buildAppButtonProps,
    buildAppRevisionButtonProps,
    buildOrgButtonProps,
    buildTagButtonProps,
    buildTagManagerButtonProps,
} from '../../../utils/BreadcrumbButtonsUtils';
import { Section, SectionProps } from '../../abstractions/Section';
import { buildAppRevisionBreadcrumbActions } from '../../../utils/BuildAppRevisionBreadcrumbActions';
import { extractPermissionsFromOrgUser } from '../../../context/OrgUserReducer';
import { SectionKey } from '../../SectionsDetails';
import { useLoggedInState } from '../../../context/AppContext';
import { useRouter } from 'next/router';
import { ChildrenAndIdProps } from '../../../types/props/ChildrenAndIdProps';
import { analyticsEnabled } from '../../../utils/AnalyticsUtils';

const TagSection: FC<ChildrenAndIdProps> = (props: ChildrenAndIdProps) => {
    const router = useRouter();

    const { id, children } = props;

    const { orgUserState, templateInteractions } = useLoggedInState();

    const { ask, dispatchDialogAction, setRefreshCurrentPage } =
        templateInteractions;
    const currentOrgPermissions = extractPermissionsFromOrgUser(orgUserState);

    const sectionProps: SectionProps<NavTag> = {
        children,
        sectionKey: SectionKey.tag,
        queryResult: useQuery<NavTag>(NavTagQuery, {
            variables: { id },
        }),
        initContext: (data) => {
            templateInteractions.setSectionHasAnalytics(analyticsEnabled(data.getTag.revision.app));
        },
        buildButtonsProps: (data) => [
            buildOrgButtonProps(
                router,
                data.me.orgs,
                data.getTag.revision.app.tag_manager_account.org.id,
                data.getTag.revision.app.tag_manager_account.org.name,
            ),
            buildTagManagerButtonProps(
                router,
                data.getTag.revision.app.tag_manager_account.id,
                data.getTag.revision.app.tag_manager_account.org.data_manager_account?.id ?? '',
            ),
            buildAppButtonProps(
                router,
                data.getTag.revision.app.tag_manager_account.apps,
                data.getTag.revision.app.id,
                data.getTag.revision.app.name,
            ),
            buildAppRevisionButtonProps(
                router,
                data.getTag.revision.app.revisions,
                data.getTag.revision.id,
                data.getTag.revision.name,
            ),
            buildTagButtonProps(router, data.getTag.revision.tags, id, data.getTag.name, true),
        ],
        buildMenuItemsProps: () => [],
        extractOrgUserDetails: (data) => data.getTag.revision.app.tag_manager_account.org,
        buildBreadcrumbActions: (data) =>
            buildAppRevisionBreadcrumbActions(
                data.getTag.revision.id,
                data.getTag.revision.name,
                router,
                currentOrgPermissions,
                dispatchDialogAction,
                setRefreshCurrentPage,
                ask,
                data.getTag.revision.locked,
            ),
        accountExpireIn: orgUserState?.tagManagerAccount?.trialExpiration ?? undefined,
        accountIsTrial: orgUserState?.tagManagerAccount?.isTrial ?? undefined,
    };

    return <Section<NavTag> {...sectionProps} />;
};

export default TagSection;
