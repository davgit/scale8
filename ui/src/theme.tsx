import { createTheme, adaptV4Theme } from '@mui/material/styles';
import { getProductSection, ProductSectionKey } from './containers/SectionsDetails';
import { grey } from '@mui/material/colors';

declare module '@mui/material/styles/createPalette' {
    interface Palette {
        adminColor: Palette['primary'];
        dataManagerColor: Palette['primary'];
        tagManagerColor: Palette['primary'];
        commonColor: Palette['primary'];
    }
    interface PaletteOptions {
        adminColor: PaletteOptions['primary'];
        dataManagerColor: PaletteOptions['primary'];
        tagManagerColor: PaletteOptions['primary'];
        commonColor: PaletteOptions['primary'];
    }
}

// A custom theme for this app
const theme = createTheme(
    adaptV4Theme({
        palette: {
            primary: {
                main: '#25302e',
            },
            secondary: {
                main: grey['700'],
            },
            error: {
                main: '#c63d51',
            },
            info: {
                main: '#444444',
            },
            success: {
                main: '#c6e54b',
            },
            background: {
                default: '#ffffff',
            },
            adminColor: {
                main: getProductSection(ProductSectionKey.admin).color,
            },
            dataManagerColor: {
                main: getProductSection(ProductSectionKey.dataManager).color,
            },
            tagManagerColor: {
                main: getProductSection(ProductSectionKey.tagManager).color,
            },
            commonColor: {
                main: getProductSection(ProductSectionKey.global).color,
            },
        },
        typography: {
            fontFamily: ['Source Sans Pro', 'Helvetica Neue', 'Arial', 'sans-serif'].join(','),
            fontSize: 14,
            body1: {
                fontSize: 14,
            },
        },
        overrides: {
            MuiTableCell: {
                head: {
                    fontWeight: 600,
                },
            },
            MuiSvgIcon: {
                root: {
                    fontSize: '1.25rem',
                },
            },
            MuiButton: {
                root: {
                    fontWeight: 600,
                },
            },
            MuiTooltip: {
                tooltip: {
                    fontSize: '.8rem',
                },
            },
            MuiDialogActions: {
                spacing: {
                    '& > :not(:first-child)': {
                        marginLeft: '16px',
                    },
                },
            },
        },
    }),
);

export default theme;
