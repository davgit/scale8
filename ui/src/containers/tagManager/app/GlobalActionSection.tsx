import { FC } from 'react';
import { useQuery } from '@apollo/client';
import { buildGlobalActionButtonProps } from '../../../utils/BreadcrumbButtonsUtils';
import { Section, SectionProps } from '../../abstractions/Section';
import NavGlobalActionQuery from '../../../gql/queries/NavGlobalActionQuery';
import { NavGlobalAction } from '../../../gql/generated/NavGlobalAction';
import { buildAppRevisionBreadcrumbActions } from '../../../utils/BuildAppRevisionBreadcrumbActions';
import { extractPermissionsFromOrgUser } from '../../../context/OrgUserReducer';
import { SectionKey } from '../../SectionsDetails';
import { useConfigState, useLoggedInState } from '../../../context/AppContext';
import { useRouter } from 'next/router';
import { ChildrenAndIdProps } from '../../../types/props/ChildrenAndIdProps';
import { analyticsEnabled, errorTrackingEnabled } from '../../../utils/AnalyticsUtils';
import { buildAppRevisionButtons } from './AppRevisionSection';

const GlobalActionSection: FC<ChildrenAndIdProps> = (props: ChildrenAndIdProps) => {
    const router = useRouter();

    const { id, children } = props;

    const { orgUserState, templateInteractions } = useLoggedInState();
    const { useSignup } = useConfigState();

    const { ask, dispatchDialogAction, setRefreshCurrentPage } = templateInteractions;
    const currentOrgPermissions = extractPermissionsFromOrgUser(orgUserState);

    const sectionProps: SectionProps<NavGlobalAction> = {
        children,
        sectionKey: SectionKey.globalAction,
        queryResult: useQuery<NavGlobalAction>(NavGlobalActionQuery, {
            variables: { id },
        }),
        sectionHasAnalytics: (data) =>
            analyticsEnabled(data.getActionGroupDistribution.revision.app),
        buildButtonsProps: (data, orgPermissions) => [
            ...buildAppRevisionButtons(
                data.me.orgs,
                data.getActionGroupDistribution.revision.app.tag_manager_account.org,
                data.getActionGroupDistribution.revision.app.tag_manager_account,
                data.getActionGroupDistribution.revision.app.tag_manager_account.org
                    .data_manager_account,
                data.getActionGroupDistribution.revision.app.tag_manager_account.apps,
                data.getActionGroupDistribution.revision.app,
                data.getActionGroupDistribution.revision.app.revisions,
                data.getActionGroupDistribution.revision,
                analyticsEnabled(data.getActionGroupDistribution.revision.app),
                errorTrackingEnabled(data.getActionGroupDistribution.revision.app),
                router,
                orgPermissions,
                useSignup,
                'Global Actions',
            ),
            buildGlobalActionButtonProps(
                router,
                data.getActionGroupDistribution.revision.global_action_group_distributions,
                id,
                data.getActionGroupDistribution.name,
                true,
            ),
        ],
        buildMenuItemsProps: () => [],
        extractOrgUserDetails: (data) =>
            data.getActionGroupDistribution.revision.app.tag_manager_account.org,
        buildBreadcrumbActions: (data) =>
            buildAppRevisionBreadcrumbActions(
                data.getActionGroupDistribution.revision.id,
                data.getActionGroupDistribution.revision.name,
                router,
                currentOrgPermissions,
                dispatchDialogAction,
                setRefreshCurrentPage,
                ask,
                data.getActionGroupDistribution.revision.locked,
            ),
        accountExpireIn: orgUserState?.tagManagerAccount.trialExpiration ?? undefined,
        accountExpired: orgUserState?.tagManagerAccount.trialExpired ?? undefined,
        accountIsTrial: orgUserState?.tagManagerAccount.isTrial ?? undefined,
    };

    return <Section<NavGlobalAction> {...sectionProps} />;
};

export default GlobalActionSection;
