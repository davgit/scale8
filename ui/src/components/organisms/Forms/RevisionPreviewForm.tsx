import { FC } from 'react';
import { DialogContent, DialogContentText } from '@mui/material';
import { FormProps } from '../../../hooks/form/useFormValidation';
import FormGqlError from '../../atoms/FormGqlError';
import DialogActionsWithCancel from '../../molecules/DialogActionsWithCancel';
import { RevisionPreviewValues } from '../../../dialogPages/tagManager/app/AppRevisionPreview';
import FormFlex from '../../atoms/FormFlex';
import { DialogFormContextProvider } from '../../../context/DialogFormContext';
import { DialogFormSelect } from '../../atoms/DialogFormInputs/DialogFormSelect';
import { DialogFormTextInput } from '../../atoms/DialogFormInputs/DialogFormTextInput';

type RevisionPreviewFormProps = FormProps<RevisionPreviewValues> & {
    environments: { name: string; url: string }[];
    handleDialogClose: (checkChanges: boolean) => void;
    description: string;
};

const RevisionPreviewForm: FC<RevisionPreviewFormProps> = (props: RevisionPreviewFormProps) => {
    return (
        <DialogFormContextProvider<RevisionPreviewValues> formProps={props}>
            <FormFlex handleSubmit={props.handleSubmit}>
                <DialogContent>
                    <DialogContentText id="alert-dialog-description">
                        {props.description}
                    </DialogContentText>
                    <FormGqlError error={props.gqlError} />
                    {props.environments.length > 0 && (
                        <DialogFormSelect
                            label="Environment"
                            name="environment"
                            values={[
                                { key: '', text: '-- None --' },
                                ...props.environments.map((_) => ({
                                    key: _.name,
                                    text: _.name,
                                })),
                            ]}
                            optional
                        />
                    )}
                    <DialogFormTextInput name="url" label="App URL" autoFocus />
                </DialogContent>
                <DialogActionsWithCancel
                    disableSubmit={props.isSubmitting}
                    handleDialogClose={props.handleDialogClose}
                    confirmText={props.submitText}
                    ignoreChanges
                />
            </FormFlex>
        </DialogFormContextProvider>
    );
};

export default RevisionPreviewForm;
