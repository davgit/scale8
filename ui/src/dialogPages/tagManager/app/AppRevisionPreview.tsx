import { Dispatch, FC, FormEvent, SetStateAction, useState } from 'react';
import { FormProps, useFormValidation } from '../../../hooks/form/useFormValidation';
import { useQuery } from '@apollo/client';
import RevisionPreviewForm from '../../../components/organisms/Forms/RevisionPreviewForm';
import { PreviewRevisionGetData } from '../../../gql/generated/PreviewRevisionGetData';
import PreviewRevisionGetQuery from '../../../gql/queries/PreviewRevisionGetQuery';
import urlValidator from '../../../utils/validators/urlValidator';
import { DialogPageProps } from '../../../types/DialogTypes';
import { usePageDialogControls } from '../../../hooks/dialog/usePageDialogControls';
import { QueryLoaderAndError } from '../../../abstractions/QueryLoaderAndError';
import RevisionPreviewDisplayForm from '../../../components/organisms/Forms/RevisionPreviewDisplayForm';

type AppRevisionPreviewAfterSubmitProps = DialogPageProps & {
    revisionName: string;
    url: string;
};

type AppRevisionPreviewAfterLoadProps = DialogPageProps & {
    revisionName: string;
    environments: { name: string; url: string }[];
};

export type RevisionPreviewValues = {
    url: string;
    environment: string;
};

const AppRevisionPreviewFormAfterSubmit: FC<AppRevisionPreviewAfterSubmitProps> = (
    props: AppRevisionPreviewAfterSubmitProps,
) => {
    const [successfullySubmitted, setSuccessfullySubmitted] = useState(false);

    const { url, revisionName, handleDialogClose } = props;

    const handleSubmit = (event?: FormEvent<HTMLFormElement>) => {
        if (event) {
            event.preventDefault();
        }

        window.open(url);
        setSuccessfullySubmitted(true);
    };

    usePageDialogControls(
        false,
        successfullySubmitted,
        props.setPageHasChanges,
        props.handleDialogClose,
        props.pageRefresh,
    );

    return (
        <RevisionPreviewDisplayForm
            handleSubmit={handleSubmit}
            submitText="Browser Preview"
            handleDialogClose={handleDialogClose}
            url={url}
            revisionName={revisionName}
        />
    );
};

const AppRevisionPreviewFormAfterLoad: FC<AppRevisionPreviewAfterLoadProps> = (
    props: AppRevisionPreviewAfterLoadProps,
) => {
    const [builtUrl, setBuiltUrl] = useState('');

    const { id, revisionName, environments } = props;

    const submitForm = async (revisionPreviewValues: RevisionPreviewValues) => {
        setBuiltUrl(`${revisionPreviewValues.url}#s8prev=${id}`);
    };

    const customValueSetter = (
        valueKey: string,
        value: any,
        values: RevisionPreviewValues,
        setValues: Dispatch<SetStateAction<RevisionPreviewValues>>,
    ) => {
        if (valueKey === 'environment') {
            const newUrl = (
                environments.find((_) => _.name === value) ?? {
                    name: '',
                    url: '',
                    installDomain: '',
                }
            ).url;

            setValues({
                ...values,
                [valueKey]: value,
                url: newUrl !== '' ? newUrl : values.url,
            });
        } else {
            setValues({
                ...values,
                [valueKey]: value,
            });
        }
    };

    const formValidationValues = useFormValidation<RevisionPreviewValues>(
        {
            url: '',
            environment: '',
        },
        [
            {
                field: 'url',
                validator: urlValidator,
                error: () => 'Invalid URL',
            },
        ],
        submitForm,
        false,
        customValueSetter,
    );

    usePageDialogControls(
        false,
        false,
        props.setPageHasChanges,
        props.handleDialogClose,
        props.pageRefresh,
    );

    if (builtUrl !== '') {
        return (
            <AppRevisionPreviewFormAfterSubmit
                {...props}
                handleDialogClose={props.handleDialogClose}
                revisionName={revisionName}
                url={builtUrl}
            />
        );
    }

    const previewDialogFormProps: FormProps<RevisionPreviewValues> = {
        ...formValidationValues,
        gqlError: undefined,
        submitText: 'Preview',
        title: 'Preview',
        handleDialogClose: props.handleDialogClose,
    };

    return (
        <RevisionPreviewForm
            {...previewDialogFormProps}
            environments={environments}
            handleDialogClose={props.handleDialogClose}
            description={`Preview Revision: ${revisionName}?`}
        />
    );
};

const AppRevisionPreview: FC<DialogPageProps> = (props: DialogPageProps) => {
    return QueryLoaderAndError<PreviewRevisionGetData>(
        false,
        useQuery<PreviewRevisionGetData>(PreviewRevisionGetQuery, {
            variables: { id: props.id },
        }),
        (data: PreviewRevisionGetData) => {
            return (
                <AppRevisionPreviewFormAfterLoad
                    revisionName={data.getRevision.name}
                    environments={data.getRevision.app.environments.map((_) => ({
                        name: _.name,
                        url: _.url ?? '',
                    }))}
                    {...props}
                />
            );
        },
    );
};

export default AppRevisionPreview;
