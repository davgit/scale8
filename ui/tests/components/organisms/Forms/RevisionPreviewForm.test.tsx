import { render } from '@testing-library/react';
import { ThemeProvider } from '@mui/material';
import theme from '../../../../src/theme';
import { FC } from 'react';
import { useMockFormProps } from '../../../componentsMocks/useMockFormProps';
import RevisionPreviewForm from '../../../../src/components/organisms/Forms/RevisionPreviewForm';
import { RevisionPreviewValues } from '../../../../src/dialogPages/tagManager/app/AppRevisionPreview';

const MockForm: FC = () => {
    const formProps = useMockFormProps<RevisionPreviewValues>({
        url: '',
        environment: '',
    });

    return (
        <RevisionPreviewForm {...formProps} description="" environments={[{ name: '', url: '' }]} />
    );
};

describe('RevisionPreviewForm', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });
    it('form render unchanged', () => {
        const { asFragment } = render(
            <ThemeProvider theme={theme}>
                <MockForm />
            </ThemeProvider>,
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
