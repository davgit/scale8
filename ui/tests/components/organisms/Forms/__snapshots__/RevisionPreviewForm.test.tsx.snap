// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RevisionPreviewForm form render unchanged 1`] = `
<DocumentFragment>
  <form
    class="MuiBox-root css-mjj15b"
  >
    <div
      class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
    >
      <p
        class="MuiDialogContentText-root MuiTypography-root MuiTypography-body1 css-s2fmnb-MuiTypography-root-MuiDialogContentText-root"
        id="alert-dialog-description"
      />
      <div
        class="MuiFormControl-root DialogFormField css-1nrlq1o-MuiFormControl-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary css-a6y1oi-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
        >
          Environment
        </label>
        <div
          class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-foxv06-MuiInputBase-root-MuiInput-root-MuiSelect-root"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="mui-component-select-environment"
            class="MuiSelect-select MuiSelect-standard MuiInput-input MuiInputBase-input css-1rxz5jq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
            id="mui-component-select-environment"
            role="button"
            tabindex="0"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            name="environment"
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-yqkyv6-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root DialogFormField css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary Mui-focused Mui-required css-1oyovms-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="mui-1"
          id="mui-1-label"
        >
          App URL
          <span
            aria-hidden="true"
            class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-b4y4zd-MuiFormLabel-asterisk"
          >
             *
          </span>
        </label>
        <div
          class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl MuiInputBase-adornedEnd css-1v2pam7-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id="mui-1"
            name="url"
            required=""
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
          >
            <button
              aria-label="toggle password visibility"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-erxxaa-MuiSvgIcon-root"
                data-testid="ClearIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiDialogActions-root MuiDialogActions-spacing css-snbeu5-MuiDialogActions-root"
    >
      <button
        class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-7kx4b6-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Cancel
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</DocumentFragment>
`;
